lock/lock.go

```diff
@@ -0,0 +1,333 @@
+// is multi-region, or global, the configuration of which regions to use should reflect that.
+func AcquireLock(options *Options) error {
+	client, err := NewDynamoDb(options.AwsRegion)
+	if err != nil {
+		options.Logger.Errorf("Error authenticating to AWS: %s\n", err)
+		return err
+	}
+
+	if options.CreateTableIfMissing == true {
+		err := createLockTableIfNecessary(options, client)
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	return acquireLockWithRetries(options, client)
+}
+
+// acquireLock will attempt to acquire the lock defined by the provided lock string in the configured lock table for the
+// configured region.
+func acquireLock(options *Options, client *dynamodb.DynamoDB) error {
+	options.Logger.Infof("Attempting to acquire lock %s in table %s in region %s\n",
+		options.LockString,
+		options.LockTable,
+		options.AwsRegion,
+	)
+
+	dynamodbSvc, err := NewDynamoDb(options.AwsRegion)
+	if err != nil {
+		options.Logger.Errorf("Error authenticating to AWS: %s\n", err)
+		return err
+	}
+
+	putParams := &dynamodb.PutItemInput{
+		Item: map[string]*dynamodb.AttributeValue{
+			attributeLockId: {S: aws.String(options.LockString)},
+		},
+		TableName:           aws.String(options.LockTable),
+		ConditionExpression: aws.String(fmt.Sprintf("attribute_not_exists(%s)", attributeLockId)),
+	}
+	_, err = dynamodbSvc.PutItem(putParams)
+	if err != nil {
+		options.Logger.Errorf(
+			"Error acquiring lock %s in table %s in region %s (already locked?): %s\n",
+			options.LockString,
+			options.LockTable,
+			options.AwsRegion,
+			err,
+		)
+		return errors.WithStackTrace(err)
+	}
+	options.Logger.Infof("Acquired lock\n")
+	return nil
+}
+
+// AcquireLockWithRetries will attempt to acquire the lock defined by the provided lock string in the configured lock table
+// for the configured region. This will retry on failure, until reaching timeout.
+func acquireLockWithRetries(options *Options, client *dynamodb.DynamoDB) error {
+	return retry.DoWithRetry(
+		options.Logger,
+		fmt.Sprintf("Trying to acquire DynamoDB lock %s at table %s\n", options.LockString, options.LockTable),
+		options.MaxRetries,
+		options.SleepBetweenRetries,
+		func() error {
+			return acquireLock(options, client)
+		},
+	)
+}
+
+// ReleaseLock will attempt to release the lock defined by the provided lock string in the configured lock table for the
+// configured region.
+func ReleaseLock(options *Options) error {
+	client, err := NewDynamoDb(options.AwsRegion)
+	if err != nil {
+		options.Logger.Errorf("Error authenticating to AWS: %s\n", err)
+		return err
+	}
+
+	tableExists, err := lockTableExistsAndIsActive(options.LockTable, client)
+	if err != nil {
+		options.Logger.Errorf("Error checking if DynamoDB table %s exists and is active\n", options.LockTable)
+		return err
+	}
+
+	if tableExists != true {
+		options.Logger.Errorf("DynamoDB table %s does not exist\n", options.LockTable)
+		return err
+	}
+
+	return retry.DoWithRetry(
```

Comment:
NIT: for consistency with `acquireLockWithRetries`, either extract this into `releaseLockWithRetries`, or, alternatively, remove `acquireLockWithRetries`, and inline that logic within `AcquireLock`.